### ISSUE TITLE
[Bug] Hunks/FilePaths memory issues

### ISSUE THREAD

**Issue by philip-iii** [SoftwareIntrospectionLab/MininGit#66]
This is somewhat related to issue 37, however the problem is elsewhere.

Hunks calls fp.update_all(repo_id) in (line 277 develop branch), which eventually causes a crash after running out of memory. update_all in turn calls update_for_revision for all commit_ids, which ultimately fails at line 128 (which as far as I gather is filling the cached adjacency matrix)

Tested with amarok git repository, on a 4GB Core 2 Duo, it fails after about 7500 revisions.

I have no idea what the solution to this could be, but I guess splitting into chunks could be one way to go about it...

**Comment by cflewis** [comment]
Yeah, I can believe this happens.

The original `cvsanaly` base doesn't store file paths in the database, but rather builds them up. We're storing a cache in memory because a large performance hit is incurred trying to search through the DB to build the paths back up again.

We're currently considering altering cvsanaly to use file paths again in the database.

A quick fix for this one would be to have a cache of, say, 1000 files, and then have old ones pushed out. I know there's at least one Python modules that will handle that for us.

**Comment by philip-iii** [comment]
I am not sure at the moment how the cache is stored, but perhaps a tree-like data structure could reduce the memory costs, Alternatively, offloading older cache entries to a temporary database, or even the entire cache could  be an option as well. Both of these options will potentially incur a performance hit and reduce the overall benefit of the caching, but will at least reduce chances of crashing.

**Comment by linzhp** [comment]
For Hunk extension, you could delete the update_all, since it builds file paths for every revision sequentially. In addition, if you are memory bounded, you could also delete the last line of FilePath.update_for_revision:

```
    self.__dict__['cached_adj'][commit_id] = deepcopy(adj)
```

which means that it won't cached any adjacency matrix. This will significantly reduce the memory usage.

**Comment by cflewis** [comment]
Cool, these are good ideas. Maybe we should have caching as a configurable option (default on).

Linzhp: Can you try changing the dictionary to a shelve object?

http://docs.python.org/library/shelve.html

This will put the dictionary on the hard disk, but it implements all dictionary methods, so all you should need to do is do the import then change the object creation. This should give us a much higher bound than memory, but still have a speed-up versus hitting the database for another query.

If you're busy with finals, let me know, and I'll do it.

**Comment by linzhp** [comment]
Configurable option is a good idea, I will do that.
However, I don't think putting the dictionary in hard disk is a good idea: disk I/O is too slow.
The original strategy in FilePath is to only cache one revision: if current revision is the same as the previous queried revision, it will not calculate the adjacency matrix again.
However, since I cached all revision, there seemed to be no need for this check, so it looks for the cache dictionary instead. I can add that check back in case the cache is turned off and adjacency matrices are built for every revision sequentially. It will work as fast as with the cache but consume far less memory.

**Comment by cflewis** [comment]
I'm not entirely sure what you mean, but I trust you know what's going on :) Looking forward to the pull request so I can see :) (comment it!)

**Comment by cflewis** [comment]
Shelf is now a possibility, but it;s not particularly stable. No more time to work on this.

**Comment by philip-iii** [comment]
I have to try out a few more things, but from initial trials, shelf now runs out of space o.O. Of course it should be a bit easier to extend than memory, but it still raises the question whether it needs to be 70+ Gigs (at the time of crash). I need to investigate this a bit further to provide more details, just mentioning for now. 

EDIT: Note that this occurred while running the BugFixMessage extension, not Hunks (which, although related, should end up in a separate issue I guess)

**Comment by cflewis** [comment]
What data set are you running it on? We've never had a problem with
space, and yet you're somehow managing to hit 70+ GBs, which is crazy
to me. We went through the entirety of the Apache httpd project (10
years of data) without a problem, so this whole patch was essentially
a shot in the dark because we can't replicate :)

I'm afraid I won't be able to work on this any more, so it might be
that you will have to either:
a) submit some magic patch ;)
b) work on a smaller data set (you might be able to chunk your
repository up into smaller pieces)
c) Patch it so that file paths aren't cached at all, and just hit the
database every single time

## 

Chris Lewis

On Mon, May 2, 2011 at 3:44 AM, philip-iii
reply@reply.github.com
wrote:

**Comment by philip-iii** [comment]
I ran it against a snapshot of the Firefox repo (converted to Git from Mercurial). I've started it over 4 hours ago and it is running at ~50GBs at the moment..

I don't know when I'll be able to look into it, cookin' on 3 burners atm, but I finally have someone who speaks Python fluently and may be able to look into this.

**Comment by cflewis** [comment]
OK, I can certainly believe Firefox could be that big :)

My suggestion is to comment out the lines in FilePaths that cache the path.
It's going to make things painful on your DB and raise your mining time, but
it should work.

Ideally, this structure should be a limited size queue which pops out the
least recently used element, so the memory size is constant. I'm sure that a
module to do that is available, but for reasons I can't really go into
publicly, we want to avoid adding dependencies. We could hide the imports
under a config flag, I guess... but we just loop back round to me not having
the time to do it right now.

If it begins to bite me in my current project, I'll update the ticket so you
know I'm working on it :)

## 

Chris Lewis (on cell phone, apologies for typos)
On May 2, 2011 8:22 AM, "philip-iii" <
reply@reply.github.com>
wrote:

**Comment by linzhp** [comment]
I am afraid commenting out caching in FilePath couldn't make Hunks work. Hunks does not use every commits in the database, though it believed it should. Last week when I am testing the shared cached_adj, I ran Hunks first and it cached matrices for every commits it reached, but when it came to run HunkBlame, there are still many commits uncached. So I ended up having to add update_all back into Hunks. Note that if you don't construct adjacent matrices for all commits in sequence, you need to call update_all, and update_all will cache all matrices.

**Comment by apepper** [comment]
Hi There.
I just wrote an e-mail to the mailinglist libresoft-tools-devel with my complainted. Just after writing that message I found this issue. So here is a copy of my e-mail:

Hello all.

I'm writing once more concerning the file paths.

Here some context: I'm running cvsanaly2 (UCSC latest version see https://github.com/SoftwareIntrospectionLab/cvsanaly) on a big project, which has a (reconstructed) git-history of over 10 years, 40000 commits, and over 75000 file entries over time.

When running cvsanaly2 with the extension Hunk or HunkBlame filepaths are needed and those extensions run "update_all". This calculates an adjacency matrix of all filenames and paths. The problem is, that this takes up HUGE amount of RAM or diskspace (using the flag Config().low_memory). In my case it took over 24 GB of RAM (diskspace using the swap file) which is all my free available space on disk. I could maybe cl es to gain more space for the swap file. The problem is that the files table tells me, that I have over 75000 file and that my size limit is already reached at not even 7000 files added to the adjacency matrix. That's not even 10% of all files. With this I would need approximately 240 GB space just to calculate the filepaths!

So my request is: I understand that you want to keep the adjacency matrix to track file movement. But could you add a second table "e.g. file_full_path" with the three columns file_id, commit_id, full_path? This table should contain for every file_id change in one commit_id the current path (at the end of the commit). With this the extension FilePaths is not needed for the extension Hunk and HunkBlame and they could perform fast mysql query (SELECT full_path FROM file_full_path WHERE file_id = ? AND commit_id = ?) and speed up dramatically.

Greetings from Berlin
Alex

**Comment by apepper** [comment]
Since I need to run cvsanaly on a large repo I wrote the needed code to do this. See https://github.com/apepper/cvsanaly/tree/new_file_path for my commits. I already added a pull-request.

### How is it working?

I created another table file_paths which contain the latest full path of a given commit and file_id. When running __action_add, __action_rename or __action_copy the path and file_id is already known and stored with the function __add_file_path all in DBContentHandler.py. So this is done, when running cvsanaly core.

Next I extended the extension FilePath with the function get_path2 and get_file_id to use the new available information from file_paths. Then using those functions from FilePath I could update FileRevs, Hunks and HunkBlame.

There are some other smaller improvements, so have a closer look at the commits, if you are interested in the details. **You need to updated https://github.com/apepper/repositoryhandler for this!**

### Benchmark

The main problem with the "old" way (using an adjacency matrix), is that it used up huge amounts of ram space (up to several GB). But it also was kind of slow, since to calculate it every file_id up to that commit_id had to be checked.

#### My setup:
- two empty mysql tables
- extensions = ['BugFixMessage', 'Patches', 'Hunks', 'HunkBlame']
- about 1000 commits and 4000 files.

#### Results:

Duration running the "old" version: 243s
Duration running the "new" version: 101s

I hope that this fixes the issue, since I really need to speed this up.

**Comment by apepper** [comment]
BTW: Feedback is welcomed, since I'm quite a novice at hacking python.

**Comment by cflewis** [comment]
Please do not merge pull request without approval from linzhp and Lewisham.
Need to review this.

Thanks for this Alexander, I'll take a look when I get to the office and
have a coffee :)

## 

Chris Lewis (on cell phone, apologies for typos)
On May 13, 2011 4:36 AM, "apepper" <
reply@reply.github.com>
wrote:

**Comment by cflewis** [comment]
Hi Alex!

I've gone through the pull request you sent to cvsanaly/. Looks great, added some comments.

Can you issue a pull request for repositoryhandler? I can do the merge on that now, I don't see anything wrong in the code.

Your benchmarks look good, and I think are around what @linzhp and I were expecting (you should try the code before @linzhp optimized it if you want to see what really slow looks like ;) ).

I'm looking forward to this!

**Comment by linzhp** [comment]
I merged the change from Alex on repositoryhandler this morning, as it looked good by itself.

**Comment by cflewis** [comment]
The patch from @apepper has been applied, but it doesn't seem to be working completely.

Try running this change at the current HEAD of develop alongside master using the [Voldemort](https://github.com/voldemort/voldemort) repo as the input.

Voldemort has a commit at `https://github.com/voldemort/voldemort/commit/fcb393f7cafe0ccca7fe42b833e996475aaf6eee` (quoted as Github is buggy with those links). None of the three files in this commit had a file path entered for them:

```
DBG: get_path_from_database: Path for file_id 5928 at commit_id 2088: None
DBG: SELECT file_path from file_paths
                   WHERE file_id=? AND commit_id <= ?
                   ORDER BY commit_id DESC LIMIT 1
DBG: get_path_from_database: Path for file_id 5928 at commit_id 2088: None
No path found for file 5928 at commit 2088
DBG: SELECT file_path from file_paths
                   WHERE file_id=? AND commit_id <= ?
                   ORDER BY commit_id DESC LIMIT 1
DBG: get_path_from_database: Path for file_id 5929 at commit_id 2088: None
DBG: SELECT file_path from file_paths
                   WHERE file_id=? AND commit_id <= ?
                   ORDER BY commit_id DESC LIMIT 1
DBG: get_path_from_database: Path for file_id 5929 at commit_id 2088: None
No path found for file 5929 at commit 2088
DBG: SELECT file_path from file_paths
                   WHERE file_id=? AND commit_id <= ?
                   ORDER BY commit_id DESC LIMIT 1
DBG: get_path_from_database: Path for file_id 5930 at commit_id 2088: None
DBG: SELECT file_path from file_paths
                   WHERE file_id=? AND commit_id <= ?
                   ORDER BY commit_id DESC LIMIT 1
DBG: get_path_from_database: Path for file_id 5930 at commit_id 2088: None
No path found for file 5930 at commit 2088
```

Because the debug output from DBContentHandler doesn't report any database IDs, I can't find the bit in the log where this was supposed to be submitted to `file_paths`. However, it is certain that we are dropping data from our extensions. That commit does exist, and the three modifications appear in the `actions` table. The file IDs are in the `files` table, so perhaps we are missing some point in the code where a file ID is created and an insert into `file_paths` doesn't happen.

Any ideas?

**Comment by cflewis** [comment]
I think I fixed this in c998cbf891beabf5c6ae12b1e8b074d77e47c4b5 on line 422 (the rest of the diff is my editor cleaning up whitespace).

It was missing an insert into file_paths.

**Comment by cflewis** [comment]
Not there yet. The other numbers (files in Content, for example) look good, but HunkBlame is very upset.

Lots of `Not a valid hunk: No previous commit found for file 3894 at commit 1321`, which means where we had 3540 hunk blames resolved, we've now dropped all the way down to 996. However, the number of hunks blamed should go up: hunks are now better linked to file IDs.

@linzhp : Can you take a look at this when you can? I know you have finals soon.

**Comment by linzhp** [comment]
OK. I will

**Comment by apepper** [comment]
any news?

**Comment by cflewis** [comment]
Unlikely, both @linzhp and I are preparing to head out to ICSE 2011, so there won't be any news on this front for a couple of weeks.

I'm 90% confident the patch is now working, otherwise I'd have done a revert (as this touches a number of files and would hold up further releases). I want to keep the ticket open until we can verify what's happening with HunkBlame, as we now have a large behavior change which needs to be documented if it turns out to be working correctly (and the old way was broken).

**Comment by linzhp** [comment]
For unknown reason, -n in "git log" does not work as expected. For example, in voldemort, 

git log --follow -n 2 --format=%H 46a5b4d7e916e3fe0a9a77659599d74709f18c2d -- .classpath

should list two revisions (if there is at least one previous revision): the previous one and the current one. However, it only listed one. While running without -n gave two revision:

git log --follow  --format=%H 46a5b4d7e916e3fe0a9a77659599d74709f18c2d -- .classpath
46a5b4d7e916e3fe0a9a77659599d74709f18c2d
fbd0f95d62ac2c5e97e5a4df5a732e9342d60da1

After removing -n from repositoryhandler, the problem fixed. Now voldemort resolves 4964 hunk blames.

**Comment by apepper** [comment]
Hi lizhp.

I couldn't reproduce your finding. When I use the flag -n 2 it yields the correct results:

$ git log --follow -n 2 --format=%H 46a5b4d7e916e3fe0a9a77659599d74709f18c2d -- .classpath
46a5b4d7e916e3fe0a9a77659599d74709f18c2d
fbd0f95d62ac2c5e97e5a4df5a732e9342d60da1
$ git log --follow  --format=%H 46a5b4d7e916e3fe0a9a77659599d74709f18c2d -- .classpath
46a5b4d7e916e3fe0a9a77659599d74709f18c2d
fbd0f95d62ac2c5e97e5a4df5a732e9342d60da1

Maybe you are running an old version of git:
`$ git --version
git version 1.7.4.4`

Could you check again, because the -n 2 falg limits output for files that have been modified very often.

Greetings from Berlin
Alex

**Comment by apepper** [comment]
Btw. There seems to be an issue using git log with --follow and -n (short for --max-count). See http://kerneltrap.org/mailarchive/git/2008/11/7/4025784 .

I'm not sure, when this is fixed, but at least with git version 1.7.4.4 it's working.

**Comment by cflewis** [comment]
FWIW, it's highly likely that the machine I tested on was also running
an old version of Git (it's a work machine that I don't have admin
rights on). I will know more later today,

## 

Chris Lewis

On Tue, May 31, 2011 at 2:09 AM, apepper
reply@reply.github.com
wrote:

**Comment by cflewis** [comment]
I'm running 1.7.3.1. Could believe problem was fixed in 1.7.4+. @linzhp, what are you running?

**Comment by linzhp** [comment]
I am running 1.7.1. Will install 1.7.4 later to check it out.

**Comment by cflewis** [comment]
Side-question: What's the `-n` used for anyway? Is it really important it's there? I think know what `-n` does from reading the docs, I'm just not sure why it's being used by repositoryhandler.

**Comment by linzhp** [comment]
Alex wants to limit the number of version returned by "git log"

**Comment by linzhp** [comment]
I installed git 1.7.5.3, the -n works now. We need to document the minimum git version to 1.7.4

**Comment by cflewis** [comment]
Alright, change the README and mark it as fixes #66.

**Comment by linzhp** [comment]
The README is updated. I added the '-n 2' back and reran HunkBlame on voldemort, getting 4741 records, comparing to 4743 without '-n 2'. Although it's not a big difference, I don't see a necessity of limiting the number of revisions returned by "git log". Even though some files might be modified very often, the list revision numbers are not likely to blow up the memory.
